services/alerts/src/AlertsManagerContext.ts

```diff
@@ -0,0 +1,62 @@
+import React from 'react'
+import { Alert, AlertsManager, AlertsManagerAlert } from './types'
+
+type SetAlertsFunction = React.Dispatch<React.SetStateAction<Alert[]>>
+
+export const makeAlertManager = (
+    setAlerts: SetAlertsFunction
+): AlertsManager => {
+    let id = 0
+
+    const remove = (alert: Alert) => {
+        setAlerts(alerts => {
+            const idx = alerts.findIndex(a => a === alert)
+
+            if (idx === -1) {
+                return alerts
+            }
+
+            return [
+                ...alerts.slice(0, idx),
+                ...alerts.slice(idx + 1, alerts.length),
+            ]
+        })
+    }
+    const add = (alert: Alert) => {
+        setAlerts(alerts => {
+            if (alerts.some(a => a === alert)) {
+                return alerts
+            }
+
+            id++
+
+            const alertManagerAlert: AlertsManagerAlert = {
+                ...alert,
+                id,
+                get remove() {
+                    return () => {
+                        alert.options.onHidden && alert.options.onHidden()
+                        remove(this)
+                    }
+                },
```

Comment:
I'm not 100% sure about this part yet... I think this makes sense, because this way you can just attach `remove` to the `onHidden` prop of the `AlertBar` and the original `options.onHidden` will be executed as well.

I also considered two alternatives:
1. Keeping `alert.remove` and `alert.options.onHidden` completely separate, i.e. not calling `alert.options.onHidden` from `alert.remove`. This would be less magical, but that would mean that the UI component would have to create an `onHidden` handler that first calls `options.onHidden` (if present) and then calls `remove`.
2. Not adding a `remove` property but redeclaring the `options.onHidden` so that it both calls the original `options.onHidden` and removes the alert from the alerts array. This would be even more magical, but it would allow the UI to simply spread the `options` as props to the `AlertBar`

Let me know your thoughts on this.

Reply:
Hmm yeah that's a tough one...

I *think* I'm actually in favor of alternative #2 - there may be cases (like when the app is unmounted from the shell on a navigation maybe?) when we'd want to remove the alert without calling `onHidden`.  Unlikely, but it's nice to have that option... we could then just use this combined function (call `onHidden` then `remove`) in the AlertStack component?

Not super opposed to the option you went with, but keeping them separate seems a bit more versatile?

Reply:
I think I'd prefer alternative 1. Less magic = easier to understand what's going on in the app, even if that means a little bit more code. But in my experience clarity is always preferable over a couple of saved LOCs

Reply:
Thanks for sharing your opinions. At first glance it seems they are opposite, but I'm not quite sure about the comment by @amcgee:
- You expressed a preference for option 2
- But in in your motivation you write _"there may be cases [...] when we'd want to remove the alert without calling onHidden"_. To me this is in line with option 1.

Please confirm.

Reply:
Sorry, I meant option 1.  Corrected

Reply:
Austin confirmed over Slack that he in fact preferred option 1.